hazelcast/src/main/java/com/hazelcast/query/impl/OrderedIndexStore.java

```diff
@@ -435,4 +480,39 @@ public Object invoke(Comparable value, Data indexKey) {
 
     }
 
+    private static final class IteratorFromBatch implements Iterator<QueryableEntry> {
+        private final Iterator<IndexKeyEntries> iterator;
+        private Iterator<QueryableEntry> indexKeyIterator;
+
+        private IteratorFromBatch(@Nonnull Iterator<IndexKeyEntries> iterator) {
+            this.iterator = iterator;
+            this.indexKeyIterator = iterator.hasNext() ? iterator.next().getEntries().iterator() : null;
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (indexKeyIterator == null) {
+                return false;
+            }
+            if (indexKeyIterator.hasNext()) {
+                return true;
+            } else {
+                if (iterator.hasNext()) {
+                    indexKeyIterator = iterator.next().getEntries().iterator();
+                    return indexKeyIterator.hasNext();
```

Comment:
If some batch has 0 entries, we will not move to the next batch. I think this is unlikely, but not impossible to happen - the batch can be a concurrent collection - maybe we remove all items from it before we remove the whole batch.

But whether it's possible or not - we should deal with this gracefully.